OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -10,6 +10,57 @@
 
 namespace LinuxTracing {
 
+bool TracerThread::OpenRingBufferForGpuTracepoint(
+    const char* tracepoint_category, const char* tracepoint_name, int32_t cpu) {
+  int fd = tracepoint_event_open(tracepoint_category, tracepoint_name, -1, cpu);
+  if (fd == -1) {
+    return false;
+  }
+  std::string buffer_name =
+      absl::StrFormat("%s:%s events", tracepoint_category, tracepoint_name);
+  PerfEventRingBuffer ring_buffer{fd, SMALL_RING_BUFFER_SIZE_KB, buffer_name};
+  if (!ring_buffer.IsOpen()) {
+    return false;
+  }
+  ring_buffers_.push_back(std::move(ring_buffer));
+
+  gpu_tracing_fds_.emplace(fd);
+  tracing_fds_.push_back(fd);
+
+  return true;
+}
+
+// This method enables events for GPU event tracing. We trace three events that
+// correspond to the following GPU driver events:
+// - A GPU job (command buffer submission) is scheduled by the application. This
+//   is tracked by the event "amdgpu_cs_ioctl".
+// - A GPU job is scheduled to run on the hardware. This is tracked by the event
+//   "amdgpu_sched_run_job".
+// - A GPU job is finished by the hardware. This is tracked by the corresponding
+//   DMA fence being signaled and is tracked by the event "dma_fence_signaled.
+// A single job execution thus correponds to three events, one of each type
+// above, that share the same timeline, context, and seqno.
+// We have to record events system-wide (per CPU) to ensure we record all
+// relevant events.
+// This method returns true on success, otherwise false.
+bool TracerThread::OpenGpuTracepoints(const std::vector<int32_t>& cpus) {
+  for (int32_t cpu : cpus) {
+    if (!OpenRingBufferForGpuTracepoint("amdgpu", "amdgpu_cs_ioctl", cpu)) {
+      return false;
```

Comment:
Yes, I think you are right. This makes this part more ugly, but I guess we have to clean up properly. 

Done.

Reply:
You use `std::vector<PerfEventRingBuffer> ring_buffers`  to store the ring buffers until you know that opening all of them succeeded, but for the file descriptor you use class member gpu_tracing_fds_. 
For consistency and to avoid filling a class member just to maybe empty it, I would have used another vector for the file descriptors and passed it to `OpenRingBufferForGpuTracepoint` as well. 
Then in `OpenGpuTracepoints` by keeping a boolean instead of returning early you could either add to tracing_fds_ gpu_tracing_fds_ and ring_buffers_, or close the file descriptors.
Not sure if more or less ugly though.

Reply:
Done.